Entity attribute setup event

Kovak opened the pull request:
The goal of this PR is to add an event where it is safe to modify LivingEntity attribute maps after registration and prior to the rest of mod loading.

At the moment it is not possible to introduce new attributes to existing entities in 1.16.x, a regression from previous versions (#6928). 

After thinking about the pitfalls of trying to hook something into the inheritance chain of static initializers of these maps, I decided it might be simpler to wait until after all registration has taken place and then just copy all the attribute maps, make them mutable again, issue the EntityAttributeSetupEvent for other mods to modify those maps, and then refreeze them afterwards.

ichttt committed:
Auto-detect ansi support for log4j2

move to fml.loading subpackage

Add license header

diesieben07 committed:
Workaround for modded attributes on vanilla entities logging warnings on vanilla clients

Switch to a handler-based approach per packet class, simplify patch, add some docs

Add license header and docs

Switch to a network handler based approach

Switch to a network handler based approach

revert build gradle change

Simplify Packet patch

Add license header to IForgePacket

Revert ForgeHooks changes

Less hacky and more generic way to filter packets for vanilla connections

_Actually_ Order SAS.
It's early

Order SAS

simplify vanilla connection check

Remove unneeded import change

Kovak committed:
initial go at an implementation for modifying entity attribute maps after registration

Kovak committed:
fix a few import changes

Kovak committed:
fix a few import changes

Kovak committed:
some style changes

Kovak committed:
some style changes

Kovak committed:
one more stylistic change

Kovak committed:
one more stylistic change

Kovak committed:
more formatting

Kovak committed:
more formatting

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=7420) <br/>All committers have signed the CLA.

Kovak committed:
rebase onto latest 1.16.x, fix typo in process attributes name

Kovak committed:
fix typo in process attributes name

cpw committed:
Rollback registries to vanilla state if something happens during loading, so subsequent activities can still run and not generate false reports..

Signed-off-by: cpw <cpw+github@weeksfamily.ca>

gigaherz committed:
Fix resource leak in the OBJ loader.

LexManos committed:
Fix mcp_snapshot mapping issue causing unneeded patches. Closes #7424

cpw committed:
Add signature reporting for mods, using new signature capture from ModLauncher. Need to figure out
how to reflect Minecraft's JAR signatures into here.

Signed-off-by: cpw <cpw+github@weeksfamily.ca>

cpw committed:
Restore FingerprintViolationEvent with appropriate THIS IS F*CKING DEAD DELET WARNINGS.
Also, notify in UI if using legacy JDK without capability to read out security data from mod jars.

Signed-off-by: cpw <cpw+github@weeksfamily.ca>

cpw committed:
Force 8.0.5 ML

Signed-off-by: cpw <cpw+github@weeksfamily.ca>

cpw committed:
Add in proper handling of equals and hashcode for modjar urls. Fixes very slow loading on windows.

Signed-off-by: cpw <cpw+github@weeksfamily.ca>

alcatrazEscapee committed:
Fix data pack biomes failing to load in SingleBiomeProvider (#7434)

sciwhiz12 committed:
Reimplement nametag distance attribute, fix attribute translation keys (#7387)

alcatrazEscapee committed:
Fix breaking changes from the Biome rename PR #7434 (#7439)

gigaherz committed:
Update forge mappings to 20201028-1.16.3 (#7435)

malte0811 committed:
Fix (NonNull)Lazy.Concurrent using a global lock (#7403)

* Fix Lazy.Concurrent using a global lock, thus preventing multiple threads from resolving independent Lazy's at the same time
* Use Lazy with an added null check to implement NonNullLazy

cpw committed:
Update modlauncher. fixes #7452

Signed-off-by: cpw <cpw+github@weeksfamily.ca>

cpw committed:
Return client resources even if no data pack found on server. Closes #6225 (Pull request)

Signed-off-by: cpw <cpw+github@weeksfamily.ca>

cpw committed:
Fix config parse failure causing crash in building exception message. fixes #7438

Signed-off-by: cpw <cpw+github@weeksfamily.ca>

cpw committed:
actually rollback on error in registry events.

Signed-off-by: cpw <cpw+github@weeksfamily.ca>

gigaherz committed:
Introduce custom loader additions to the model data generators. (#7450)

Currently implemented loaders:
* OBJ
* Composite
* Multi-layer
* Item layers (vanilla item/generated but with fullbright texture support)
* Bucket
* Separate Perspective

LexManos committed:
Fix Biomes not properly copying over ids loaded from the save to the new Dynamic Registry.

LexManos committed:
1.16.4 Initial Update

cpw committed:
Fix early sorting bug that meant the file deduping was causing an error, to, well, not cause it, so we can get to an error screen.

Signed-off-by: cpw <cpw+github@weeksfamily.ca>

cpw committed:
Update event bus version (should be 3.0.5)

Tweak MDK to automatically populate recommended values for things like forge version, loader version and minecraft version.

Signed-off-by: cpw <cpw+github@weeksfamily.ca>

gigaherz committed:
Add FluidStack codec, and a test mod that verifies its behaviour matches the existing write/read logic.
Add missing license headers.

cpw committed:
Support backwards loading 1.16.3 mods in 1.16.4, because we are able to do that. Tweak loading a bit to be smarter about dependency errors of various kinds.

Signed-off-by: cpw <cpw+github@weeksfamily.ca>

cpw committed:
Handle erroring during early mod construction phases, and actually report that into the error UI
by doing armslength exception handling and propagation.

Signed-off-by: cpw <cpw+github@weeksfamily.ca>

cpw committed:
Some help in ExtensionPoint regarding DISPLAYTEST. Let's improve those tick rates!

Signed-off-by: cpw <cpw+github@weeksfamily.ca>

Kovak committed:
change to the mod loading stage api, fix mods.toml

Kovak committed:
update patch for some of the name changes, fix the ModLoadingStage to properly use the new 4th arg

Kovak committed:
merge head
